pkg/cluster/px/px.go

```diff
@@ -20,26 +39,199 @@ type Cluster interface {
+	if err != nil {
+		logrus.Errorf("failed updating daemonset: %s. Err: %s", ds.Name, err)
+		return err
+	}*/
 	return nil
 }
 
-func (p *pxCluster) Destroy(c *apiv1alpha1.Cluster) error {
-	logrus.Infof("destroying px cluster: %s", c.Name)
+func (ops *pxClusterOps) Destroy(c *apiv1alpha1.Cluster) error {
+	logrus.Warnf("destroying px cluster is implemented")
 	return nil
 }
 
-// NewPXClusterProvider creates a new PX cluster
-func NewPXClusterProvider(conf interface{}) (Cluster, error) {
-	return &pxCluster{}, nil
+// runDockerPuller runs the DaemonSet to start pulling the given image on all nodes
+func (ops *pxClusterOps) runDockerPuller(imageToPull string) error {
+	pullerName := "docker-puller"
+
+	labels := map[string]string{
+		"name": pullerName,
+	}
+	args := []string{"-i", imageToPull, "-w"}
+
+	var env []corev1.EnvVar
+	if len(ops.dockerRegistrySecret) != 0 {
+		logrus.Infof("using user-provided docker registry credentials from secret: %s", ops.dockerRegistrySecret)
+		env = append(env, corev1.EnvVar{
+			Name: "REGISTRY_USER",
+			ValueFrom: &corev1.EnvVarSource{
+				SecretKeyRef: &corev1.SecretKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: ops.dockerRegistrySecret,
+					},
+					Key: "REGISTRY_USER",
+				},
+			},
+		})
+
+		env = append(env, corev1.EnvVar{
+			Name: "REGISTRY_PASS",
+			ValueFrom: &corev1.EnvVarSource{
+				SecretKeyRef: &corev1.SecretKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: ops.dockerRegistrySecret,
+					},
+					Key: "REGISTRY_PASS",
+				},
+			},
+		})
+	}
+
+	ds := &extensionsv1beta1.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      pullerName,
+			Namespace: pxDefaultNamespace,
+			Labels:    labels,
+		},
+		Spec: extensionsv1beta1.DaemonSetSpec{
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: labels,
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:            pullerName,
+							Image:           dockerPullerImage,
+							ImagePullPolicy: corev1.PullAlways,
+							Args:            args,
+							Env:             env,
+							ReadinessProbe: &corev1.Probe{
+								InitialDelaySeconds: 60,
+								Handler: corev1.Handler{
+									Exec: &corev1.ExecAction{
+										Command: []string{"cat", "/tmp/docker-pull-done"},
+									},
+								},
+							},
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      "varrun",
+									MountPath: "/var/run/",
+								},
+							},
+						},
+					},
+					RestartPolicy: "Always",
+					Volumes: []corev1.Volume{
+						{
+							Name: "varrun",
+							VolumeSource: corev1.VolumeSource{
+								HostPath: &corev1.HostPathVolumeSource{
+									Path: "/var/run/",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	policy := metav1.DeletePropagationForeground
+	err := ops.kubeClient.Extensions().DaemonSets(pxDefaultNamespace).Delete(
```

Comment:
Should comment on the reason why we need to delete first. Is this to account for running only one daemonset if multiple requests come?